index.js

```diff
@@ -1,174 +1,255 @@
+
+  // connection timeout logic
+  let tid
+  let timeoutHit = false
+  if (this.options.connectionTimeoutMillis) {
+    tid = setTimeout(() => {
+      this.log('ending client due to timeout')
+      timeoutHit = true
+      // force kill the node driver, and let libpq do its teardown
+      client.connection ? client.connection.stream.destroy() : client.end()
+    }, this.options.connectionTimeoutMillis)
+  }
 
-      client.release = function (err) {
-        delete client.release
-        if (err) {
-          this.log('destroy client. error:', err)
-          this.pool.destroy(client)
-        } else {
-          this.log('release client')
-          this.pool.release(client)
-        }
-      }.bind(this)
-
-      if (cb) {
-        cb(null, client, client.release)
+  this.log('connecting new client')
+  client.connect((err) => {
+    this.log('new client connected')
+    if (tid) {
+      clearTimeout(tid)
+    }
+    client.on('error', idleListener)
+    if (err) {
+      // remove the dead client from our list of clients
+      this._clients = this._clients.filter(c => c !== client)
+      if (timeoutHit) {
+        err.message = 'Connection terminiated due to connection timeout'
+      }
+      cb(err, undefined, NOOP)
+    } else {
+      client.release = release.bind(this, client)
+      this.emit('connect', client)
+      this.emit('acquire', client)
+      if (this.options.verify) {
+        this.options.verify(client, cb)
       } else {
-        resolve(client)
+        cb(undefined, client, client.release)
       }
-    }.bind(this))
-  }.bind(this))
+    }
+  })
+  return result
 }
 
-Pool.prototype.take = Pool.prototype.connect
-
 Pool.prototype.query = function (text, values, cb) {
   if (typeof values === 'function') {
     cb = values
     values = undefined
   }
-
-  return this._promise(cb, function (resolve, reject) {
-    this.connect(function (err, client, done) {
+  const response = this._promisify(cb)
+  cb = response.callback
+  this.connect((err, client) => {
+    if (err) {
+      return cb(err)
+    }
+    this.log('dispatching query')
+    client.query(text, values, (err, res) => {
+      this.log('query dispatched')
+      client.release(err)
       if (err) {
-        return reject(err)
+        return cb(err)
+      } else {
+        return cb(undefined, res)
       }
-      client.query(text, values, function (err, res) {
-        done(err)
-        err ? reject(err) : resolve(res)
-      })
     })
-  }.bind(this))
+  })
+  return response.result
 }
 
 Pool.prototype.end = function (cb) {
-  this.log('draining pool')
-  return this._promise(cb, function (resolve, reject) {
-    this.pool.drain(function () {
-      this.log('pool drained, calling destroy all now')
-      this.pool.destroyAllNow(resolve)
-    }.bind(this))
-  }.bind(this))
+  this.log('ending')
+  if (this.ending) {
+    const err = new Error('Called end on pool more than once')
+    return cb ? cb(err) : this.Promise.reject(err)
```

Comment:
In general, Its not good to mix async and sync in the same API. Whats weird here is nothing invoked internally is actually async except the result. In one case I can get the error synchronously (looks async) and the other, async.

If I were to express a preference it would be to stick with an async api throughout.

Here is a tree of posts that go into a lot more detail http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony